### PR TITLE
only filter out the entire parent subset if it has any partitions being materialized this tick

### PR DESCRIPTION
**Pull request by gibsondan**
Summary:
Try to make the fix in https://github.com/dagster-io/dagster/pull/31315 even more targeted without affecting throughput as much - if a parent is actively being materialized this tick, be more careful about what child partitions get submitted, to minimize the risk of inadvertently combining them into a single run incorrectly. But if we are just reacting to parent materializations and the parent itself is not being materialized on this tick too, only remove the specific partitions that are

> Insert changelog entry or delete this section.

## Changelog

> Insert changelog entry or delete this section.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 1]
```diff
@@ -1529,7 +1529,7 @@ def _execute_asset_backfill_iteration_inner(
         candidate_asset_graph_subset = _get_candidate_asset_graph_subset(
             asset_backfill_data,
             asset_graph_view,
-            materialized_since_last_tick,
+            updated_materialized_subset,
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py]: what's the reasoning for this change? this makes it so we re-evaluate all assets that have a parent that has ever been materialized every tick, which in theory is somewhat unrelated to the other changes.

ideally, this would be fine (i.e. one possibility if that we just evaluate everything all the time), but there's some rough edge cases in the self-dependent asset tests that made me a bit scared of that sort of change.

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 2]
```diff
@@ -1672,6 +1672,13 @@ def _should_backfill_atomic_asset_subset_unit(
         )
         entity_subset_to_filter = entity_subset_to_filter.compute_difference(requested_partitions)
 
+    any_parent_being_requested_this_tick_subset = any(
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py]: nit: odd name here -- should it just be `any_parent_being_requested_this_tick` given that it's a boolean?

**Review by OwenKephart — COMMENTED** [review]
generally aligned on the approach, just commenting for now because some things might need to change to get all tests to pass

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py]: to add a bit more color (as I just looked at bk and saw that a similar test was failing), the core issue here is that the code for self-dependent assets _implicitly depends on_ the fact that we only evaluate the asset when one of its parents has updated this tick.

this is because our logic doesn't actually currently have any provisions forcing the upstream partition to be materialized once it's being evaluated. so if we evaluate the whole self-dependent asset every time, each evaluation will emit a batch of partitions for the self-dependent asset.

**Comment by gibsondan** [comment]
Ah ok - the thinking was that we shouldn’t need  to wait for the upstream
asset to materialize necessarily, if the only reason the downstream asset
didn’t execute was because there was a risk of it inadvertently causing a
lineage violation - it could happen on the very next tick in theory even
before anything materializes

**Comment by gibsondan** [comment]
reverted the materialized_since_last_tick => updated_materialized_subset change

**Comment by gibsondan** [comment]
* **#31381** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31381?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#31461](https://github.com/dagster-io/dagster/pull/31461) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31461?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#31477](https://github.com/dagster-io/dagster/pull/31477) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31477?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31381?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#31430** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31405** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31405?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31403** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31403?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31450** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31450?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by gibsondan** [comment]
### Merge activity

* **Jul 31, 4:55 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31381).
* **Jul 31, 5:07 PM UTC**: [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31381) rebased this pull request as part of a merge.
* **Jul 31, 5:09 PM UTC**: [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31381) rebased this pull request as part of a merge.
* **Jul 31, 5:10 PM UTC**: @gibsondan merged this pull request with [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31381).
